### ISSUE TITLE
Failed to import the Cloud Storage client library for Node.js

### ISSUE THREAD

**Issue by truonggiangdao** [firebase/firebase-tools#1384]
Please help me, I really tried to google search for the whole evening and still can't find a good solution for it.

### [REQUIRED] Environment info

**firebase-tools:** 6.11.0

**Platform:** Windows

**Node version on my machine: ** v8.11.2

### [REQUIRED] Test case
I just want to do a `admin.storage().bucket(myFirebaseStorageBucket)`  so that I can upload file to my firebase project storage. But I can't. `firebase serve` will throw me following error once I try to do that:

`Error: Failed to import the Cloud Storage client library for Node.js. Make sure to install the "@google-cloud/storage" npm package. Original error: TypeError: Cannot read property '0' of undefined`
 
Check my package.json:
```
{
  "name": "xpersity-api",
  "description": "Rest API for Xpersity",
  "version": "0.1.0",
  "scripts": {
    "lint": "eslint . --fix",
    "start": "firebase serve --debug",
    "logs": "firebase functions:log",
    "shell": "firebase functions:shell",
    "deploy": "firebase deploy --only functions",
    "deploy:index": "firebase deploy --only firestore:indexes"
  },
  "engines": {
    "node": "8"
  },
  "husky": {
    "hooks": {
      "pre-commit": "npm run lint",
      "pre-push": "npm run lint"
    }
  },
  "dependencies": {
    "@google-cloud/storage": "^2.5.0",
    "axios": "^0.18.0",
    "body-parser": "^1.19.0",
    "cookie-parser": "^1.4.4",
    "cors": "^2.8.5",
    "crypto": "^1.0.1",
    "custom-env": "^1.0.1",
    "dotenv": "^8.0.0",
    "express": "^4.17.1",
    "express-validation": "^1.0.2",
    "firebase-admin": "~7.0.0",
    "firebase-functions": "^2.3.0",
    "helmet": "^3.18.0",
    "http-status": "^1.3.2",
    "joi": "^14.3.1",
    "method-override": "^3.0.0",
    "mkdirp": "^0.5.1",
    "mkdirp-promise": "^5.0.1",
    "moment": "^2.24.0",
    "morgan": "^1.9.1",
    "nodemailer": "^6.2.1",
    "query-string": "^6.5.0",
    "require-context": "^1.1.0",
    "simple-oauth2": "^2.2.1",
    "uuid": "^3.3.2"
  },
  "devDependencies": {
    "cpy-cli": "^2.0.0",
    "eslint": "^5.12.0",
    "eslint-plugin-promise": "^4.0.1",
    "firebase-functions-test": "^0.1.6",
    "husky": "^2.3.0",
    "nodemon": "^1.19.1"
  }
}
```

### [REQUIRED] Steps to reproduce

First I initializeApp:
```
const firebaseStorageBucket = 'my-project-dev-abc.appspot.com';
admin.initializeApp({
  credential: admin.credential.cert(serviceAccount),
  databaseURL: config.firestoreDbUrl,
  storageBucket: firebaseStorageBucket ,
});
```

`serviceAccount` is json data I download from my project's firebase console. Using this I can already access firebase firestore. Now I want to save a file into storage:

```
const bucket = () => (
  admin.storage().bucket(firebaseStorageBucket)
);
```

and then when I try to use `bucket().upload` the same with samples in `firebase/functions-samples/convert-images` => I got the exception saying that I must install the "@google-cloud/storage" npm package. But I already did. And already try to test on other Windows machine as well. But still the same issue.

### [REQUIRED] Expected behavior

I checked in my `node_modules` already have the `@google-cloud/storage` package, so base on the sample it should work. If I just commented out the part about `bucket().upload` then the rest of my code work normally. I don't know where I did wrong.

### [REQUIRED] Actual behavior

Here my debug log:

```
i  Your code has been provided a "firebase-admin" instance.
>  Finished initialize app.
i  functions: HTTP trigger initialized at http://localhost:5001/my-project-dev-abc/us-central1/api
[hosting] Rewriting /api/expert/register to http://localhost:5001/my-project-dev-abc/us-central1/api for local Function api
[2019-06-10T20:10:21.561Z] Accepted request POST /my-project-dev-abc/us-central1/api/api/expert/register --> api
[2019-06-10T20:10:22.700Z] Functions runtime initialized.
i  functions: Beginning execution of "api"
[2019-06-10T20:10:22.703Z] Disabled runtime features: {"functions_config_helper":true,"network_filtering":true,"timeout":true,"memory_limiting":true,"protect_env":true,"admin_stubs":true}
[2019-06-10T20:10:25.872Z] firebase-admin has been stubbed.
i  Your code has been provided a "firebase-admin" instance.
>  Finished initialize app.
[2019-06-10T20:10:27.653Z] Trigger "api" has been found, beginning invocation!
[2019-06-10T20:10:27.654Z]
[2019-06-10T20:10:27.655Z] Running api in mode HTTPS
[2019-06-10T20:10:27.672Z] {"socketPath":"\\\\?\\pipe\\D:\\api\\functions\\6484"}
[2019-06-10T20:10:27.680Z] [functions] Runtime ready! Sending request! {"socketPath":"\\\\?\\pipe\\D:\\api\\functions\\6484"}
[2019-06-10T20:10:27.790Z] Ephemeral server used!
[2019-06-10T20:10:28.383Z] Ephemeral server survived.
>  uploadAvatar.error: { Error: Failed to import the Cloud Storage client library for Node.js. Make sure to install the "@google-cloud/storage" npm package. Original error: TypeError: Cannot read property '0' of undefined
>      at new FirebaseError (D:\api\functions\node_modules\firebase-admin\lib\utils\error.js:42:28)
>      at new Storage (D:\api\functions\node_modules\firebase-admin\lib\storage\storage.js:101:19)
>      at D:\api\functions\node_modules\firebase-admin\lib\firebase-app.js:254:20
>      at FirebaseApp.ensureService_ (D:\api\functions\node_modules\firebase-admin\lib\firebase-app.js:339:23)
>      at FirebaseApp.storage (D:\api\functions\node_modules\firebase-admin\lib\firebase-app.js:252:21)
>      at Proxy.fn (D:\api\functions\node_modules\firebase-admin\lib\firebase-namespace.js:311:45)
>      at bucket (D:\api\functions\app\common\helpers\firebase\bucket.helper.js:46:34)
>      at Object.uploadAvatar (D:\api\functions\app\common\helpers\file.helper.js:112:11)
>      at <anonymous>
>    errorInfo:
>     { code: 'storage/missing-dependencies',
>       message: 'Failed to import the Cloud Storage client library for Node.js. Make sure to install the "@google-cloud/storage" npm package. Original error: TypeError: Cannot read property \'0\' of undefined' } }
>  profile.avatar null
127.0.0.1 - - [10/Jun/2019:20:10:30 +0000] "POST /api/expert/register HTTP/1.1" 200 53 "-" "PostmanRuntime/7.13.0"
>  POST /api/expert/register 200 2189.905 ms - 53
i  functions: Finished "api" in ~2s
```

let me know if I can provide further details,
Thanks

**Comment by Memeriaj** [comment]
I think that you're doing the correct thing here, but I'm not completely certain.  Have you tried deleting your `npm_modules` directory and then `npm install`-ing it again?

It seems that `@google-cloud/storage` is an [optional dependency in the `firebase-admin-node` repo](https://github.com/firebase/firebase-admin-node/blob/master/package.json#L66).  I had thought that NPM would try installing all of the optional dependencies and then just not error out the overall install if one of them fails (normally this would be due to platform specific requirements).  @hiranya911 do you have any suggestions?

**Comment by truonggiangdao** [comment]
hi @Memeriaj thanks, I already did that and also tried to test on other machine as well. It still show the same message. And I also able to deploy to `firebase functions` and test, but again the same issue.

**Comment by truonggiangdao** [comment]
**Update**: I still don't know what cause this issue in my project, but I did some more tests and found out that if I create a `firebase functions` without `express` then the issue not occur. So for now my workaround is I have another repo for my second functions which only do the upload to `Storage` and my current app just send api to it if need upload.

**Comment by abeisgoat** [comment]
If you go into your `functions` directory and run...

```
node -e "console.log(require.resolve('@google-cloud/storage'))"` what do you get?

**Comment by truonggiangdao** [comment]
hi guys, thanks for support me troubleshooting, I finally figured out the cause of this issue. It because I used package [require-context](https://www.npmjs.com/package/require-context) on my project and it changes the way `require` work. I just remove the package from my project and then everything working normally.

It has nothing todo with `firebase-tools`. For that I will close the ticket. Thanks!
